mtt-management/src/mtt_types.rs

```diff
@@ -0,0 +1,446 @@
+fn set_l1_entry(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let l2_entry = get_l2_entry(phys_addr.bits());
+    match l2_entry {
+        MttL1Dir(pfn) => update_l1_value(pfn, phys_addr, memory_type),
+        _ => unreachable!(),
+    }
+}
+
+/// Changes the L2 type for 16 consecutive entries (each representing a 64MB range) for the 1GB range
+/// containing the address to the specified type.
+/// This can be called to consolidate previously fractured ranges.
+pub fn set_1gb_range_type(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let range_start = PageSize::Size1G.round_up(phys_addr.bits());
+    for i in 0u64..=15 {
+        let addr = range_start + (i * 64 * 1024 * 1024);
+        let mtt_entry = get_l2_entry(addr);
+        if matches!(mtt_entry, MttInvalidL2) {
+            panic!("Invalid MTT entry for address {addr:x}");
+        }
+        match memory_type {
+            MttMemoryType::NonConfidential => set_l2_entry(addr, MttNonConfidential1G),
+            MttMemoryType::Confidential => set_l2_entry(addr, MttConfidential1G),
+        }
+    }
+}
+
+// Changes the L2 type of each of the 16 consecutive entries for the the 1GB range containing the address
+// from 1GB to 2MBx32MB.
+// All pages in the range will retain the original MTT type (non-confidential or confidential)
+fn fracture_1gb_range(phys_addr: SupervisorPageAddr) {
+    let range_start = PageSize::Size1G.round_up(phys_addr.bits());
+    let first_entry_type = get_l2_entry(range_start);
+    // Enforce the invariant that all 16 L2 entries for the 1GB range must have the same type
+    for i in 0u64..=15 {
+        let addr = range_start + (i * 64 * 1024 * 1024);
+        let mtt_entry = get_l2_entry(addr);
+        match mtt_entry {
+            MttNonConfidential1G if mtt_entry == first_entry_type => {
+                set_l2_entry(addr, MttL2EntryType::non_confidential_64_mb_range())
+            }
+            MttConfidential1G if mtt_entry == first_entry_type => {
+                set_l2_entry(addr, MttL2EntryType::confidential_64_mb_range())
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+// Changes the L2 type of 64MB range containing the address to 4Kx16KB.
+// All pages in the range will retain the original MTT type (non-confidential or confidential).
+// The `mtt_l1_page` is supplied by the caller and will contain 2-bit encoding for every page
+// in the fractured range. The caller must ensure that the page isn't freed until there are
+// no outstanding references to it in the L2 table.
+fn fracture_64mb_range(phys_addr: SupervisorPageAddr, mtt_l1_page: SupervisorPageAddr) {
+    let addr = mtt_l1_page.bits();
+    // Safety: mtt_l1_page is a supervisor page address that's owned by us, and is at least 4K in size
+    unsafe {
+        core::ptr::write_bytes(addr as *mut u8, 0, PageSize::Size4k as usize);
```

Comment:
addr will always be 4KB aligned?

Reply:
Yes, by virtue of phys_addr being of type SupervisorPageAddr.